src/compile/scale.ts

```diff
@@ -43,8 +51,46 @@ export function compileScales(channels: Channel[], model: Model) {
         }
       });
 
-      return scaleDef;
-    });
+      // add additional scales needed to support ordinal color scales also for legends
+      if (channel === COLOR && fieldDef.legend && (fieldDef.type === ORDINAL || fieldDef.bin || fieldDef.timeUnit)) {
+        scales.push({
+          name: INVERSE_RANK,
+          type: ORDINAL,
+          domain: {
+            data: model.dataTable(),
+            // use rank_<field> for ordinal type, for bin and timeUnit use default field
+            field: model.field(COLOR, (fieldDef.bin || fieldDef.timeUnit) ? {} : {prefn: 'rank_'}), sort: true
+          },
+          range: {data: model.dataTable(), field: model.field(COLOR), sort: true}
```

Comment:
Looks like you never use INVERSE_RANK for bin and timeUnit case except in legend.ts  but I don't think it is necessary.